### PR TITLE
Restrict IPv6 Access to IMDS

### PR DESCRIPTION
**Pull request by hanwen-pcluste**
Signed-off-by: Hanwen <hanwenli@amazon.com>

### Description of changes
In addition to lockdown 169.254.169.254, this PR locks down fd00:ec2::254.

These two IP addresses are used to retrieve instance metedata.

Note that even prior to this PR, clusters created by ParalllelCluster do not support ipv6 IMDS access. However, users can enable it by: 1. Assigning an IPv6 address to the instances. 2. Runing `aws ec2 modify-instance-metadata-options --instance-id <instance-id> --http-protocol-ipv6 enabled`.

After this commit, unauthorized users do not have access to ipv6 IMDS even if it is enabled with the steps above.

This lockdown can be turn on/off in cluster configuration file.

### Tests
This commit is manually tested:
1. Created a new instance with the new cookbook.
2. Enabled the IMDS IPv6 access by: a. Assigning an IPv6 address to the instances. b. Run `aws ec2 modify-instance-metadata-options --instance-id <instance-id> --http-protocol-ipv6 enabled`.
3. Verify `curl http://[fd00:ec2::254]/latest/meta-data/` is denied

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [x] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cookbooks/aws-parallelcluster-config/files/default/imds/imds-access.sh [hunk 1]
```diff
@@ -116,47 +120,54 @@ main() {
...
-    info "Deleting rules related to IMDS access for user: $user"
-    iptables_delete $PARALLELCLUSTER_CHAIN $IMDS_IP "ACCEPT" $user
-    iptables_delete $PARALLELCLUSTER_CHAIN $IMDS_IP "REJECT" $user
-  done
-
-  # Add rule: ACCEPT user, for every allowed user
-  for user in $allow_users; do
-    info "Allowing IMDS access for user: $user"
-    iptables_add $PARALLELCLUSTER_CHAIN $IMDS_IP "ACCEPT" $user
+  for ip_version_address in iptables,$IMDS_IP ip6tables,$IMDS_IP6
```
Comment (francesco-giordano) [inline cookbooks/aws-parallelcluster-config/files/default/imds/imds-access.sh]: Minor: I'ld put the 2 iptables command as constant
```bash
IPTABLE_CMD_V4="iptables"
IPTABLE_CMD_V6="ip6tables"
```

**Review by francesco-giordano — COMMENTED** [review]
If we do not install explicitly `iptable` and `ip6tables`, what about add a little kitchen to test the existence of the 2 binaries. Some OSs might remove them in future version

File: cookbooks/aws-parallelcluster-config/files/default/imds/imds-access.sh (hunk 1 above)
Comment (hanwen-pcluste) [reply cookbooks/aws-parallelcluster-config/files/default/imds/imds-access.sh]: Done

**Comment by hanwen-pcluste** [comment]
> If we do not install explicitly `iptable` and `ip6tables`, what about add a little kitchen to test the existence of the 2 binaries. Some OSs might remove them in future version

Many of our integration tests will be able to detect the issue. Because the script will fail if any of the commands does not exist.

**Review by francesco-giordano — APPROVED** [review]
LGTM!

File: cookbooks/aws-parallelcluster-config/files/default/imds/imds-access.sh [hunk 2]
```diff
@@ -35,10 +35,11 @@ EOF
 }
 
 function iptables_delete() {
-  local chain=$1
-  local destination=$2
-  local jump=$3
-  local user=$4
+  local $iptables_command=$1
```
Comment (enrico-usai) [inline cookbooks/aws-parallelcluster-config/files/default/imds/imds-access.sh:38]: I think there is an error here. We shouldn't have the $ symbol on the left

Comment (hanwen-pcluste) [reply cookbooks/aws-parallelcluster-config/files/default/imds/imds-access.sh:38]: Thanks! https://github.com/aws/aws-parallelcluster-cookbook/pull/1464 fixes this bug

**Pull request merged**